### PR TITLE
[JENKINS-72449] Specify that no fallback to the default locale should be used when looking up a resource bundle via `I18n` action.

### PR DESCRIPTION
**Pull request by Vlatombe**
This action is used afaik in the context of setup wizard. It does not impact classic messages resolution.
Blue Ocean is also affected, however it uses its own action, also affected by the same problem.

When running the JVM with a default locale that is not english, the resource bundle lookup for english would return a bundle with that default locale, instead of using the "default" that is english.

Also changed bundle resolution to use uberClassloader rather than iterating on all plugin classloaders.

I kept the cache usage for json representation of resource bundle, even if I think its interest is questionable, resource bundles being already cached in their object form by the JDK.

See [JENKINS-72449](https://issues.jenkins.io/browse/JENKINS-72449).

### Testing done

* Launched `java -Duser.language=fr -jar jenkins.war` in a fresh `JENKINS_HOME`. With a browser only configured in english, verified that the setup wizard is displayed exclusively in english. Before this change, part of the strings would be in french.

### Proposed changelog entries

- JENKINS-72449, display strings consistently in the requested language when running Jenkins in a JVM with a non-english locale.

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [X] The Jira issue, if it exists, is well-described.
- [X] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [X] There is automated testing or an explanation as to why this change has no tests.
- [X] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [X] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/jenkins/util/ResourceBundleUtil.java
Comment (jglick) [inline core/src/main/java/jenkins/util/ResourceBundleUtil.java:69]: I guess there is no style policy in core beyond what little Checkstyle enforces, but is there any precedent for using `var`?

**Review by jglick — APPROVED** [review]
Not entirely sure I follow the nature of the bug (some sort of cache poisoning I guess) but anyway this seems to be used only by https://github.com/jenkinsci/jenkins/blob/24dd6c2180e0ba9544914a297da8e911ba61b9ba/core/src/main/java/jenkins/I18n.java#L57-L110 so the impact should be limited.

**Comment by Vlatombe** [comment]
> some sort of cache poisoning

This is not related to cache, this is due to the fact the default bundles are in english.

If I have default JVM locale set to `fr`, and I request bundle `Messages` in `en`, the logic is the following:
* Does `Messages_en.properties` exist? No
* Does `Messages.properties` exist? Yes, this is the root bundle.
If the only bundle it finds for a given language is the root bundle, then the default `ResourceBundle.Control` implementation falls back to the default locale.
* Does `Message_fr.properties` exist ? Yes (in my example)
Then it returns a bundle composed of `fr` strings, defaulting to the root bundle for strings not defined in the `fr` bundle.

Passing `ResourceBundle.Control.getNoFallbackControl(ResourceBundle.Control.FORMAT_DEFAULT)` instead disables explicitly this fallback mechanism to the default locale.

Messages in Jenkins using localizer have their own custom logic, and checking it, I found this comment https://github.com/kohsuke/localizer/blob/master/lib/src/main/java/org/jvnet/localizer/ResourceBundleHolder.java#L102-L104 showing the same issue was hit, but did not find about `ResourceBundle.Control.getNoFallbackControl`

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Review by NotMyFault — APPROVED** [review]

**Comment by basil** [comment]
The **Testing Done** section of the PR template is mandatory, not optional.

**Pull request merged**
